pkg/streamer/reader.go

```diff
@@ -103,9 +117,203 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
 	return nil
 }
 
-// StartSync start syncon
+// getUUIDByGTID gets uuid subdir which contain the gtid set
+func (r *BinlogReader) getUUIDByGTID(gset mysql.GTIDSet) (string, error) {
+	// TODO: use a better mechanism to call relay.meta.Flush
+	// get the meta save in memory
+	relayMetaHub := GetRelayMetaHub()
+	relayMeta := relayMetaHub.GetMeta()
+
+	// get flush logs from oldest to newest
+	for _, uuid := range r.uuids {
+		filename := path.Join(r.cfg.RelayDir, uuid, utils.MetaFilename)
+		fd, err := os.Open(filename)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		defer fd.Close()
+
+		var meta Meta
+		_, err = toml.DecodeReader(fd, &meta)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+
+		gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, meta.BinlogGTID)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		if gs.Contain(gset) {
+			r.tctx.L().Info("get uuid subdir by gtid", zap.Stringer("GTID Set", gset), zap.String("uuid", uuid))
+			return uuid, nil
+		}
+	}
+
+	// use memory meta
+	if len(relayMeta.UUID) > 0 {
+		gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, relayMeta.BinlogGTID)
+		if err != nil {
+			return "", terror.ErrRelayLoadMetaData.Delegate(err)
+		}
+		if gs.Contain(gset) {
+			r.tctx.L().Info("get uuid subdir by gtid", zap.Stringer("GTID Set", gset), zap.String("uuid", relayMeta.UUID))
+			return relayMeta.UUID, nil
+		}
+	}
+	return "", terror.ErrNoUUIDDirMatchGTID.Generate(gset.String())
+}
+
+// GetFilePosByGTID tries to get Pos by GTID for file
+func (r *BinlogReader) GetFilePosByGTID(ctx context.Context, filePath string, gset mysql.GTIDSet) (uint32, error) {
+	ctx2, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s := newLocalStreamer()
+	parser := replication.NewBinlogParser()
+	parser.SetVerifyChecksum(true)
+	parser.SetUseDecimal(true)
+	if r.cfg.Timezone != nil {
+		parser.SetTimestampStringLocation(r.cfg.Timezone)
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		offset := int64(4)
+		onEventFunc := func(e *replication.BinlogEvent) error {
+			offset = int64(e.Header.LogPos)
+
+			select {
+			case s.ch <- e:
+			case <-ctx2.Done():
+			}
+			return nil
+		}
+
+		err := parser.ParseFile(filePath, offset, onEventFunc)
```

Comment:
seems this `ParseFile` will open file, read some content and parse, seek, and close. maybe we could use `ParseSingleEvent` or something to reuse last parser internal status

Reply:
We need to parse the whole file until we meet first previous_gtid_event/xid_event/query_event which not contain gset. If use ParseSingleEvent, we still need these logic(seek/close...)

Reply:
I wrongly read this file, thought it's in a for-loop 😢